### ISSUE TITLE
facilitate composition of our low level functional API to reduce end user friction

### ISSUE THREAD

**Issue by tomwayson** [Esri/arcgis-rest-js#339]
One area of friction with our functional approach is that sometimes multiple requests that are sequential or at least made w/in the same context will share parameters that will need passed into each fn call. For example, when working w/ a feature service, you may want to call [queryFeatures](https://esri.github.io/arcgis-rest-js/api/feature-service/queryFeatures/), [getFeature](https://esri.github.io/arcgis-rest-js/api/feature-service/queryFeatures/), and the [getAttachments](https://esri.github.io/arcgis-rest-js/api/feature-service/getAttachments/) in succession, all w/ the same `url` (and maybe the same `authentication`).

An OO solution would be to add `new FeatureService({ url })` that would expose methods that wrap the stateless functions passing in `this.url` to each. I'm pretty sure none of us (least of all @dbouwman)  want to get into that business due to the risk of memory leaks etc associated w/`new`ing up instances in browsers.

Instead, I propose we provide at least an example somewhere of how to use functional techniques like higer order functions to compose our low level functional API and reduce the above friction. For example: 

```js
import { queryFeatures, getFeature } from '@esri/arcgis-rest-feature-service';

// create functions to query the trees service
const treesUrl = 'https://services.arcgis.com/V6ZHFr6zdgNZuVG0/arcgis/rest/services/Landscape_Trees/FeatureServer/0';
const queryTrees = withUrl(treesUrl, queryFeatures);
const getTree = withUrl(treesUrl, getFeature);
```

Here's a working JSBin w/ what I have in mind: https://jsbin.com/geravah/edit?html,js,console

I think the question for me is where to include this information? A guide topic w/ a linked JSBin? Or would we go so far to include common helpers like `withUrl()` and/or `withAuthentication()` in the libraries? If so, in what package?

**Comment by tomwayson** [comment]
Or do we only provide a `withOptions(commonRequestOptions, fn)` that would work at least on any fn that has only a single requestOptions argument?

**Comment by dbouwman** [comment]
100% agree with this sentiment and approach. In `hub.js` I've put functional helpers in the `hub-common` module, but have not added higher order fn's like this yet. 

Perhaps put the `withAuthentication` in the `-authentication` package...
And I think the `withUrl` seems to fit w/ the feature-service package as that's really where we'd be manipulating urls

**Comment by tomwayson** [comment]
Thanks @dbouwman.

If we go w/ individual fns, putting each in the packages you suggest makes a lot of sense, and if you wanted to apply more than you could compose them like:

```js
import { UserSession, withAuthentication } from '@esri/arcgis-rest-auth';
import { queryFeatures, withUrl } from '@esri/arcgis-rest-feature-service';

const userSession = new UserSession({ /*  */ });

const queryTreesAsUser = withUrl(treesUrl, withAuthentication(userSession, queryFeatures));
```
What's nice about that is that the composing fns live in packages that the consuming app is already `import`ing from.

That said, I'm starting to realize there might be a few of these, for example maybe a `withPortalUrl()` which I think would have to live in `-request`. 

That's why I'm leaning towards at least starting w/ a composing fn that could live in `-request` like:

```js
import { UserSession } from '@esri/arcgis-rest-auth';
import { queryFeatures } from '@esri/arcgis-rest-feature-service';
import { withRequestOptions } from '@esri/arcgis-rest-request';

const userSession = new UserSession({ /*  */ });

const queryTreesAsUser = withRequestOptions({
  authentication: userSession,
  url: treesUrl
}, queryFetures);
```

Especially if we want to handle the fn's that don't just take a single argument that is the request options - it'd be better to contain and test that logic in one place.

To me `withRequestOptions()` is the first PR.

We could then create the other componsing fns using that one:

```ts
// in @esri/arcgis-rest-auth/src/with-authentication.ts
import { withRequestOptions } from '@esri/arcgis-rest-request';

export function withAuthentication (authentication, fn) {
  return withRequestOptions({
    authentication
  }, fn);
}
```

I think that gives users the most flexibility.

**Comment by jgravois** [comment]
`withOptions()` and `setDefaultOptions()` will be included in `v2.0.0`
